README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->
![Static Badge](https://img.shields.io/badge/flutter__gradient__animation__text-v1.0.3-8a2b) ![Static Badge](https://img.shields.io/badge/nullsafety-3.0.0-00b7eb) ![Static Badge](https://img.shields.io/badge/platform-flutter-ff69b4.svg) ![Static Badge](https://img.shields.io/badge/license-MIT-purple)


This package lets you make animated gradient texts without any hassle. 
Just pass the ```text```, ```colors```, the ```duration``` of the animation and you're done!
Using the ```transform``` and ```reverse```, you can achieve amazing effects.
For More info read the docs below.

## Getting started

Install the package using the following command:
```bash
flutter pub add flutter_gradient_animation_text
```
Import the package in your dart file using the following code:
```dart
import 'package:flutter_gradient_animation_text/flutter_gradient_animation_text.dart';
```
And you are good to go!

## Usage

Create your animated gradient text.
The 3 required arguments are:
1. ```text```(Text): The text.
2. ```colors```(List\<Colors>): List of colors you want in your gradient.
3. ```duration```(Duration): Duration for which a single gradient animation runs.
```dart
// rainbow text
GradientAnimationText(
    text: Text(
        'Rainbow Text 1',
        style: TextStyle(
            fontSize: 50,
        ),
    ),
    colors: [
        Color(0xff8f00ff),  // violet
        Colors.indigo,
        Colors.blue,
        Colors.green,
        Colors.yellow,
        Colors.orange,
        Colors.red,
    ],
    duration: Duration(seconds: 5),
),
```
![Rainbow Text 1][rainbow-text-1]

Reverse the animation.
This plays the animation backwards when a single iteration of the forward animation is completed.
You set ```reverse``` to true to enable this.
```dart
// rainbow text with reverse
GradientAnimationText(
    text: Text(
        'Rainbow Text 1',
        style: TextStyle(
            fontSize: 50,
        ),
    ),
    colors: [
        Color(0xff8f00ff),  // violet
        Colors.indigo,
        Colors.blue,
        Colors.green,
        Colors.yellow,
        Colors.orange,
        Colors.red,
    ],
    duration: Duration(seconds: 5),
    reverse: true,  // reverse
),
```
![Rainbow Text 2][rainbow-text-2]

Play with colors.
It is possible to use the same color more than once to increase the area it covers.
Gold Text Example:
The following 2 Gold Text animations have different number of ```Colors.amber``` color.
```dart
// gold text 1
GradientAnimationText(
    text: Text(
        'Gold Text 1',
        style: TextStyle(
            fontSize: 50,
            fontWeight: FontWeight.bold,
        ),
    ),
    colors: [
        Colors.amber,   // 1
        Colors.white,
    ],
    duration: Duration(seconds: 5),
    reverse: true,
),
// gold text 2
GradientAnimationText(
    text: Text(
        'Gold Text 2',
        style: TextStyle(
            fontSize: 50,
            fontWeight: FontWeight.bold,
        ),
    ),
    colors: [
        Colors.amber,   // 1
        Colors.amber,   // 2
        Colors.amber,   // 3
        Colors.white,
    ],
    duration: Duration(seconds: 5),
    reverse: true,
),
```
![Gold Text][gold-text]

Transform your gradient.
You can transform your gradient using ```transform``` property.
Use [GradientTranform][gradient-transform] classes such as [SweepGradient][sweep-gradient], and [GradientRotation][gradient-rotation] to tranform.
```dart
// Tranform Text
GradientAnimationText(
    text: Text(
    'Tranform Text',
        style: TextStyle(
            fontSize: 90,
            fontWeight: FontWeight.bold,
        ),
    ),
    colors: [
        Color(0xFF061A9C),
        Color(0xff92effd),
    ],
    duration: Duration(seconds: 5),
    transform: GradientRotation(math.pi / 4),   // tranform
),
```
![Tranform Text][transform-text]

## Available Properties
| Property        	| Type               	| Description                                          	|
|-----------------	|--------------------	|------------------------------------------------------	|
| ```text```      	| Text               	| The text to be used in the animation.                	|
| ```colors```    	| List<Color>        	| Colors of the animation.                             	|
| ```duration```  	| Duration           	| Duration for which a single gradient animation runs. 	|
| ```reverse```   	| bool?              	| If you want to reverse the animation.                	|
| ```transform``` 	| GradientTransform? 	| You can transform your gradient.                     	|
## Author
This package is developed by Sahil Shah.

- [GitHub][github]
- [LinkedIn][linkedin]

[rainbow-text-1]: https://github.com/SahilDShaw/flutter-gradient-animation-text/raw/2ebd0eed8db6f220ce2eb628caca880f0594a201/images/rainbow-text-1.gif

[rainbow-text-2]: https://github.com/SahilDShaw/flutter-gradient-animation-text/raw/2ebd0eed8db6f220ce2eb628caca880f0594a201/images/rainbow-text-2.gif

[gold-text]: https://github.com/SahilDShaw/flutter-gradient-animation-text/raw/2ebd0eed8db6f220ce2eb628caca880f0594a201/images/gold-text.gif

[transform-text]: https://github.com/SahilDShaw/flutter-gradient-animation-text/raw/2ebd0eed8db6f220ce2eb628caca880f0594a201/images/transform-text.gif

[gradient-transform]: https://api.flutter.dev/flutter/painting/GradientTransform-class.html

[sweep-gradient]: https://api.flutter.dev/flutter/painting/SweepGradient-class.html

[gradient-rotation]: https://api.flutter.dev/flutter/painting/GradientRotation-class.html

[github]: https://github.com/SahilDShaw

[linkedin]: https://www.linkedin.com/in/sahil-shah-pro/


CHANGELOG.md
## 1.0.0

* The initial release.

## 1.0.1
* Unnecessary dependencies removed.

## 1.0.2
* Documentation updated.

test/flutter_gradient_animation_text_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_gradient_animation_text/flutter_gradient_animation_text.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/example.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import 'package:flutter_gradient_animation_text/flutter_gradient_animation_text.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Gradient Text',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(
        title: const Text('Flutter Gradient Text'),
      ),
      body: const Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            // Rainbow Text Without Reverse
            GradientAnimationText(
              text: Text(
                'Rainbow Text 1',
                style: TextStyle(
                  fontSize: 50,
                ),
              ),
              colors: [
                Color(0xff8f00ff),
                Colors.indigo,
                Colors.blue,
                Colors.green,
                Colors.yellow,
                Colors.orange,
                Colors.red,
              ],
              duration: Duration(seconds: 5),
            ),
            // Rainbow Text With Reverse
            GradientAnimationText(
              text: Text(
                'Rainbow Text 1',
                style: TextStyle(
                  fontSize: 50,
                ),
              ),
              colors: [
                Color(0xff8f00ff),
                Colors.indigo,
                Colors.blue,
                Colors.green,
                Colors.yellow,
                Colors.orange,
                Colors.red,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Gold Text 1
            GradientAnimationText(
              text: Text(
                'Gold Text 1',
                style: TextStyle(
                  fontSize: 50,
                  fontWeight: FontWeight.bold,
                ),
              ),
              colors: [
                Colors.amber,
                Colors.white,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Gold Text 2
            GradientAnimationText(
              text: Text(
                'Gold Text 2',
                style: TextStyle(
                  fontSize: 50,
                  fontWeight: FontWeight.bold,
                ),
              ),
              colors: [
                Colors.amber,
                Colors.amber,
                Colors.amber,
                Colors.white,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Tranform Text
            GradientAnimationText(
              text: Text(
                'Tranform Text',
                style: TextStyle(
                  fontSize: 90,
                  fontWeight: FontWeight.bold,
                ),
              ),
              colors: [
                Color(0xFF061A9C),
                Color(0xff92effd),
              ],
              duration: Duration(seconds: 5),
              transform: GradientRotation(math.pi / 4),
            ),
          ],
        ),
      ),
    );
  }
}


lib/flutter_gradient_animation_text.dart
library flutter_gradient_animation_text;

import 'package:flutter/material.dart';

class GradientAnimationText extends StatefulWidget {
  // data members
  final Text text;
  final List<Color> colors;
  final Duration duration;
  final bool? reverse;
  final GradientTransform? transform;

  // constructor
  const GradientAnimationText({
    required this.text,
    required this.colors,
    required this.duration,
    this.reverse,
    this.transform,
    super.key,
  });

  @override
  State<GradientAnimationText> createState() => _GradientAnimationTextState();
}

class _GradientAnimationTextState extends State<GradientAnimationText> with SingleTickerProviderStateMixin {
  // data members
  late AnimationController _animationController;
  late Animation _animation;
  late List<Color> colors;
  late int n;
  late double diff;
  late Text text;

  @override
  void initState() {
    super.initState();
    text = widget.text;
    // animation controller
    _animationController = AnimationController(
      vsync: this,
      duration: widget.duration,
    );

    // repeating animation
    _animationController.repeat(reverse: widget.reverse ?? false);

    // animation
    _animation = Tween(begin: 0.0, end: 1.0).animate(_animationController)
      ..addListener(() {
        setState(() {});
      });

    // colors
    colors = [];
    colors.add(widget.colors.last);
    colors.addAll(widget.colors);
    colors.addAll(widget.colors);

    // colors length
    n = widget.colors.length;

    // diff
    diff = (1 / n);
  }

  // creates stops dynamically based on the animation value
  List<double> stopsList() {
    int multiplier = -1 * n;
    List<double> stops = [];

    // generates (2*n + 1) stops
    while (multiplier <= n) {
      stops.add(_animation.value + (multiplier * diff));
      multiplier++;
    }

    return stops;
  }

  @override
  void dispose() {
    // disposing animationController
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ShaderMask(
      // copying the text with color as Colors.white
      child: Text(
        text.data!,
        key: text.key,
        locale: text.locale,
        maxLines: text.maxLines,
        overflow: text.overflow,
        textAlign: text.textAlign,
        selectionColor: text.selectionColor,
        semanticsLabel: text.semanticsLabel,
        softWrap: text.softWrap,
        strutStyle: text.strutStyle,
        style: (text.style != null) ? text.style!.copyWith(color: Colors.white) : const TextStyle(color: Colors.white),
        textDirection: text.textDirection,
        textHeightBehavior: text.textHeightBehavior,
        textScaleFactor: text.textScaleFactor,
        textWidthBasis: text.textWidthBasis,
      ),
      shaderCallback: (rect) {
        // gradient
        return LinearGradient(
          tileMode: TileMode.clamp,
          transform: widget.transform,
          stops: stopsList(),
          colors: colors,
        ).createShader(rect);
      },
    );
  }
}


example/lib/main.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import 'package:flutter_gradient_animation_text/flutter_gradient_animation_text.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Gradient Text',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(
        title: const Text('Flutter Gradient Text'),
      ),
      body: const Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            // Rainbow Text Without Reverse
            GradientAnimationText(
              text: Text(
                'Rainbow Text 1',
                style: TextStyle(
                  fontSize: 50,
                ),
              ),
              colors: [
                Color(0xff8f00ff),
                Colors.indigo,
                Colors.blue,
                Colors.green,
                Colors.yellow,
                Colors.orange,
                Colors.red,
              ],
              duration: Duration(seconds: 5),
            ),
            // Rainbow Text With Reverse
            GradientAnimationText(
              text: Text(
                'Rainbow Text 1',
                style: TextStyle(
                  fontSize: 50,
                ),
              ),
              colors: [
                Color(0xff8f00ff),
                Colors.indigo,
                Colors.blue,
                Colors.green,
                Colors.yellow,
                Colors.orange,
                Colors.red,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Gold Text 1
            GradientAnimationText(
              text: Text(
                'Gold Text 1',
                style: TextStyle(
                  fontSize: 50,
                  fontWeight: FontWeight.bold,
                ),
              ),
              colors: [
                Colors.amber,
                Colors.white,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Gold Text 2
            GradientAnimationText(
              text: Text(
                'Gold Text 2',
                style: TextStyle(
                  fontSize: 50,
                  fontWeight: FontWeight.bold,
                ),
              ),
              colors: [
                Colors.amber,
                Colors.amber,
                Colors.amber,
                Colors.white,
              ],
              duration: Duration(seconds: 5),
              reverse: true,
            ),
            // Tranform Text
            GradientAnimationText(
              text: Text(
                'Tranform Text',
                style: TextStyle(
                  fontSize: 90,
                  fontWeight: FontWeight.bold,
                ),
                textAlign: TextAlign.center,
              ),
              colors: [
                Color(0xFF061A9C),
                Color(0xff92effd),
              ],
              duration: Duration(seconds: 5),
              transform: GradientRotation(math.pi / 4),
            ),
          ],
        ),
      ),
    );
  }
}
